### PR TITLE
[feature/whats new screen] In-App changelog "What´s new screen"

### PR DESCRIPTION
**Pull request by hosy**
## Description
If the user has auto update enabled on the iOS device, the user cannot see the AppStore changelog, what changed in the new app version.

With this new screen the user can see an In-App changelog, when the app was started the first time, after an app update.

If there was more than one app update since the last app launch, the user will see everything what's new, since the last started app version.

It is important, that this file `ownCloud/Release Notes/ReleaseNotes.plist` needs new content before an update was submitted to Apple.
This file also contains the image data as a base64 encoded string.

The `ReleaseNotesHostViewController` can later easily used to integrate an `Intro` or `Onbording` view.

## Motivation and Context
Better marketing to the user.

## How Has This Been Tested?
- Install app version < 1.2
- Launch app
- Install app version 1.2 
- Launch app `What´s new screen` should appear
- Force quite App 
- Launch app again `What´s new screen` should NOT appear

## Screenshots (if appropriate):
Simulator Screen Shot - iPhone 11 Pro - 2019-12-08 at 11 52 12

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING**](https://github.com/owncloud/ios-app/blob/master/CONTRIBUTING.md) document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

## QA

Test plan:

Bugs & improvements:

- [X] (1) Showing only after upgrading https://github.com/owncloud/ios-app/pull/572#issuecomment-572457884  [DONE]
- [x] (2) Link to App Store https://github.com/owncloud/ios-app/pull/572#issuecomment-572460624 [DONE]
- [x] (3) After killing the app, what's new is there again https://github.com/owncloud/ios-app/pull/572#issuecomment-572461273 [FIXED]

### CODE REVIEW

File: ownCloud/Release Notes/ReleaseNotesHostViewController.swift [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+			headerView.heightAnchor.constraint(equalToConstant: headerHeight)
+		])
+
+		titleLabel.translatesAutoresizingMaskIntoConstraints = false
+		titleLabel.setContentHuggingPriority(UILayoutPriority.defaultLow, for: NSLayoutConstraint.Axis.horizontal)
+
+		titleLabel.text = "New in ownCloud".localized
+		titleLabel.textAlignment = .center
+		titleLabel.numberOfLines = 0
+		titleLabel.font = UIFont.systemFont(ofSize: UIFont.systemFontSize * 1.5, weight: .bold)
```
Comment (mneuwert) [inline ownCloud/Release Notes/ReleaseNotesHostViewController.swift]: @hosy can we use dynamic type font for this? Something like `.headline` style or similar.

File: ownCloud/Release Notes/ReleaseNotesHostViewController.swift [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+			view.addSubview(bottomView)
+			NSLayoutConstraint.activate([
+				bottomView.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor),
+				bottomView.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor),
+				bottomView.topAnchor.constraint(equalTo: containerView.bottomAnchor),
+				bottomView.bottomAnchor.constraint(equalTo: view.safeAreaLayoutGuide.bottomAnchor)
+			])
+
+			proceedButton.setTitle("Proceed".localized, for: .normal)
+			proceedButton.layer.cornerRadius = cornerRadius
```
Comment (mneuwert) [inline ownCloud/Release Notes/ReleaseNotesHostViewController.swift]: @hosy I noticed that we use different corner radius values for buttons, sometimes 8.0 and sometimes 10.0. Could we unify it and e.g. provide some category on `UIButton` with some factory method or similar?

File: ownCloud/Release Notes/ReleaseNotesHostViewController.swift [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+			proceedButton.translatesAutoresizingMaskIntoConstraints = false
+			proceedButton.addTarget(self, action: #selector(dismissView), for: .touchUpInside)
+			bottomView.addSubview(proceedButton)
+
+			footerLabel.textAlignment = .center
+			footerLabel.translatesAutoresizingMaskIntoConstraints = false
+			footerLabel.text = "Thank you for using ownCloud.\nIf you like our App, please leave an AppStore review.\n❤️".localized
+			footerLabel.numberOfLines = 0
+			footerLabel.font = UIFont.systemFont(ofSize: 14.0)
```
Comment (mneuwert) [inline ownCloud/Release Notes/ReleaseNotesHostViewController.swift]: @hosy Another case for dynamic type?

File: ownCloud/Release Notes/ReleaseNotesHostViewController.swift [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+			}
+
+			return false
+		}
+
+		return true
+	}
+
+	func releaseNotes(for version: String) -> [[String:Any]]? {
+		if let path = Bundle.main.path(forResource: "ReleaseNotes", ofType: "plist") {
```
Comment (mneuwert) [inline ownCloud/Release Notes/ReleaseNotesHostViewController.swift]: @hosy Looking at code in lines 154-172, I think there is some potential to reduce code duplication...

File: ownCloud/Release Notes/ReleaseNotesHostViewController.swift (hunk 1 above)
Comment (hosy) [reply ownCloud/Release Notes/ReleaseNotesHostViewController.swift]: fixed

Comment (hosy) [reply ownCloud/Release Notes/ReleaseNotesHostViewController.swift]: fixed

File: ownCloud/Release Notes/ReleaseNotesHostViewController.swift (hunk 2 above)
Comment (hosy) [reply ownCloud/Release Notes/ReleaseNotesHostViewController.swift]: now using ThemeButton

File: ownCloud/Release Notes/ReleaseNotesHostViewController.swift (hunk 3 above)
Comment (hosy) [reply ownCloud/Release Notes/ReleaseNotesHostViewController.swift]: fixed

File: ownCloud/Release Notes/ReleaseNotesHostViewController.swift (hunk 4 above)
Comment (hosy) [reply ownCloud/Release Notes/ReleaseNotesHostViewController.swift]: removed one if statement

**Comment by jesmrec** [comment]
Changed base branch to `milestone/1.3`

**Comment by jesmrec** [comment]
### (1) improvement

I'd set the view only in case of upgrading to newest version, so that user checks the new features after upgrading.

In case of a fresh install, i'd not show this. For a new user everything in the app is new, not only the latest released features. As you mentioned in the first message, for new users i'd set a welcome wizard or onboarding/welcome view in a later issue.

what do you think?

**Comment by jesmrec** [comment]
### (2) (improvement)

A link to the app in App Store would be helpful here:

Screenshot 2020-01-09 at 10 01 18

**Comment by jesmrec** [comment]
### (3)

1. Install from scratch -> What's new is there
2. Kill the app
3. Open again

Current: What's new is displayed again
Expected: What's new is displayed only once

iPadPro v13.2
iPhoneXR v12

**Comment by hosy** [comment]
@jesmrec (1), (2), (3) was fixed by my side. Request for app rating is now integrated. Changed also some release notes strings.

File: ownCloud/Release Notes/ReleaseNotesHostViewController.swift [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+			headerView.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor),
+			headerView.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor),
+			headerView.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor),
+			headerView.heightAnchor.constraint(equalToConstant: headerHeight)
+		])
+
+		titleLabel.translatesAutoresizingMaskIntoConstraints = false
+		titleLabel.setContentHuggingPriority(UILayoutPriority.defaultLow, for: NSLayoutConstraint.Axis.horizontal)
+
+		titleLabel.text = "New in ownCloud".localized
```
Comment (jesmrec) [inline ownCloud/Release Notes/ReleaseNotesHostViewController.swift]: `ownCloud` word should be brandable

File: ownCloud/Release Notes/ReleaseNotesHostViewController.swift [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+				bottomView.topAnchor.constraint(equalTo: containerView.bottomAnchor),
+				bottomView.bottomAnchor.constraint(equalTo: view.safeAreaLayoutGuide.bottomAnchor)
+			])
+
+			proceedButton.setTitle("Proceed".localized, for: .normal)
+			proceedButton.translatesAutoresizingMaskIntoConstraints = false
+			proceedButton.addTarget(self, action: #selector(dismissView), for: .touchUpInside)
+			bottomView.addSubview(proceedButton)
+
+			footerButton.setTitle("Thank you for using ownCloud.\nIf you like our App, please leave an AppStore review.\n❤️".localized, for: .normal)
```
Comment (jesmrec) [inline ownCloud/Release Notes/ReleaseNotesHostViewController.swift]: `ownCloud` word should be brandable

File: ownCloud/Resources/en.lproj/Localizable.strings [hunk 7]
```diff
@@ -489,6 +489,11 @@
 "Save File" = "Save File";
 "Choose an account and folder to import the file into.\n\nOnly one file can be imported at once." = "Choose an account and folder to import the file into.\n\nOnly one file can be imported at once.";
 
+/* Release Notes */
+"Proceed" = "Proceed";
+"New in ownCloud" = "New in ownCloud";
+"Thank you for using ownCloud.\nIf you like our App, please leave an AppStore review.\n❤️" = "Thank you for using ownCloud.\nIf you like our App, please leave an AppStore review.\n❤️";
```
Comment (jesmrec) [inline ownCloud/Resources/en.lproj/Localizable.strings]: `ownCloud` word should be brandable

**Comment by jesmrec** [comment]
I keep having problems to get the `what's new`, so i have some questions to clarify.

The typical scenario i perform is:

1. installing the current stable version `1.2.1` in a device , checking in `Settings` that everything is correct.

Screenshot 2020-01-14 at 17 27 43


2. Go to XCode -> ownCloud  target -> General -> Version, and set a higher version and build number.  (1.3 as version, and incrementing in one the build number)

Screenshot 2020-01-14 at 17 25 28


3. Run the app to upgrade it in the same device, and check the `Settings`to see there the new app version and build number (they match correctly)

Screenshot 2020-01-14 at 17 31 38


I tried with and without the beta warning, but the `what's new` is not there in anycase. 

By checking the code, here is where you check if the release notes should be displayed:

https://github.com/owncloud/ios-app/blob/db724f597afeef800434d1b27da870dd5c57efcd/ownCloud/Release%20Notes/ReleaseNotesHostViewController.swift#L154

I set there a break point, and the `if` clause in the next row always returns false, so the block returns `false` (tell me if i'm wrong)

Specific questions:

- Is it correct to upgrade only both fields mentioned in step 2.? anything else to do?
- ipa installed must be an archived release ipa? or is it enough with debug ipa?
- Does Beta warning anything to do with the upgrading? (i guess not)

maybe i'm missing something. I hope these details help.

**Comment by hosy** [comment]
@jesmrec please check again

**Comment by jesmrec** [comment]
Latest round with commit `6ee498d`

(1) -> `What's new` is displayed when the app is installed from srcratch

(2), (3) -> done!

**Comment by hosy** [comment]
@jesmrec I noticed that `.isBetaBuild` always exists, even if the app was installed from the scratch. The only solution to prevent showing this screen after a fresh install is, to check, if there is existing an account.
I cannot catch the following scenario:
- User installed version 1.2.1
- No account is configured in the app
- User updating the app to version 1.3, with no account configured
- App will be started
- `What's new View` does not appear

This will not happen, if an account was configured before updating, but I cannot catch this case.
Please check this with the latest commit 6af51be

**Comment by jesmrec** [comment]
It's no problem to assume that restriction. If almost one account is attached, it's working fine now.

Approved.

**Comment by hosy** [comment]
@jesmrec app name is now customizable in this view

**Comment by jesmrec** [comment]
Approved then.

**Pull request merged**